### PR TITLE
nawk: 20180827 -> unstable-2021-02-15, fix noweb and nawk cross-complation

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
Fix cross-compilation of noweb and nawk. Test with

```ShellSession
$ nix build .#pkgsCross.raspberryPi.noweb
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 114657` at 37214f27 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/b9449397956f8455fbe50389e7e7d4d7">nawk</a></li></li>
    <li><a href="https://gist.github.com/00c37cd33698a48f4e533a5af5325881">noweb</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/development/tools/literate-programming/noweb/default.nix:14:15:
```
   |
14 |   patches = [ ./no-FAQ.patch ];
   |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/text/nawk/default.nix:21:3:
```
   |
21 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `targetPackages`.
Near pkgs/development/tools/literate-programming/noweb/default.nix:1:64:
```
  |
1 | { lib, stdenv, fetchFromGitHub, nawk, groff, icon-lang ? null, targetPackages }:
  |                                                                ^
```
</li>
  </ul>
</details>

File: pkgs/development/tools/literate-programming/noweb/default.nix [hunk 1]
```diff
@@ -11,25 +11,24 @@ let noweb = stdenv.mkDerivation rec {
...
+    # Remove FAQ
+    ./no-FAQ.patch
+  ];
 
   nativeBuildInputs = [ groff ] ++ lib.optionals (!isNull icon-lang) [ icon-lang ];
+  buildInputs = [ nawk ];
 
   preBuild = ''
     mkdir -p "$out/lib/noweb"
     cd src
```
Comment (SuperSandro2000) [inline pkgs/development/tools/literate-programming/noweb/default.nix:24]: Should we use sourceRoot instead?

File: pkgs/development/tools/literate-programming/noweb/default.nix [hunk 2]
```diff
@@ -11,25 +11,24 @@ let noweb = stdenv.mkDerivation rec {
...
+    substituteInPlace Makefile --replace 'strip' '${stdenv.cc.targetPrefix}strip'
```
Comment (SuperSandro2000) [inline pkgs/development/tools/literate-programming/noweb/default.nix:25]: This belongs more into postPatch.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 114657` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>nawk</li>
<li>noweb</li>
</ul>
</details>

**Comment by siraben** [comment]
Any blockers?

**Review by yurrriq — APPROVED** [review]
lgtm, thanks!

**Comment by siraben** [comment]
@yurrriq by the way, why does `noweb` need to be a fixpoint?

**Comment by yurrriq** [comment]
It's for `passthru.pkgs = [ noweb.tex ]` which is so users can add noweb to their TeX Live stacks, e.g. `texlive.combine { inherit noweb; }`. I'm open to better approaches.

File: pkgs/development/tools/literate-programming/noweb/default.nix [hunk 3]
```diff
@@ -11,25 +11,28 @@ let noweb = stdenv.mkDerivation rec {
...
+    # Remove FAQ
+    ./no-FAQ.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace Makefile --replace 'strip' '${stdenv.cc.targetPrefix}strip'
+  '';
 
   nativeBuildInputs = [ groff ] ++ lib.optionals (!isNull icon-lang) [ icon-lang ];
+  buildInputs = [ nawk ];
```
Comment (yurrriq) [inline pkgs/development/tools/literate-programming/noweb/default.nix:26]: Can we do this or similar? If noweb is built with Icon, we don't need nawk at runtime.

```suggestion
  buildInputs = lib.optionals (isNull icon-lang) [ nawk ];
```

File: pkgs/development/tools/literate-programming/noweb/default.nix [hunk 4]
```diff
@@ -54,7 +57,7 @@ let noweb = stdenv.mkDerivation rec {
         # NOTE: substituteInPlace breaks Icon binaries, so make sure the script
         #       uses (n)awk before calling.
         if grep -q nawk "$f"; then
-            substituteInPlace "$f" --replace "nawk" "${gawk}/bin/awk"
+            substituteInPlace "$f" --replace "nawk" "${nawk}/bin/awk"
```
Comment (yurrriq) [inline pkgs/development/tools/literate-programming/noweb/default.nix:60]: Will you please add `$out/lib/noweb/pipedocs` to this of files to patch?

Comment (siraben) [reply pkgs/development/tools/literate-programming/noweb/default.nix:60]: Ok. Was it added between the two versions?

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 114657` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>nawk</li>
    <li>noweb</li>
  </ul>
</details>

File: pkgs/development/tools/literate-programming/noweb/default.nix [hunk 5]
```diff
@@ -1,6 +1,6 @@
-{ lib, stdenv, fetchFromGitHub, gawk, groff, icon-lang ? null }:
+{ lib, stdenv, fetchFromGitHub, nawk, groff, icon-lang ? null }:
```
Comment (SuperSandro2000) [inline pkgs/development/tools/literate-programming/noweb/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, nawk, groff, icon-lang }:
```

Comment (yurrriq) [reply pkgs/development/tools/literate-programming/noweb/default.nix:1]: `icon-lang ? null` was put in place so the package builds with (n)awk by default, and Icon if provided. Are you suggesting to make Icon the default?

File: pkgs/development/tools/literate-programming/noweb/default.nix (hunk 4 above)
Comment (yurrriq) [reply pkgs/development/tools/literate-programming/noweb/default.nix:60]: Thanks!

**Comment by yurrriq** [comment]
has someone tested this with `icon-lang = null`?

**Comment by SuperSandro2000** [comment]
> has someone tested this with `icon-lang = null`?

overriding packages to null is something we try to not introduce anymore. Boolean flags are preferred.

**Comment by yurrriq** [comment]
> overriding packages to null is something we try to not introduce anymore. Boolean flags are preferred.

That's all fine and good, but this package still works that way. Thanks for checking.

**Pull request merged**